### PR TITLE
Feature/make_gurkan

### LINKED ISSUE #32: Automatically give Gurkan role on name change
This feature would have the bot check a member's name every time it changes and if a variant of gurkan is in the name, it would give the gurkan role, else if it is not, it would remove the role. This would be done using the "on_member_update" function of discord.py.

### PR DESCRIPTION
**Pull request by RandomDev26**
Closes #32

### CODE REVIEW

File: bot/exts/gurkan/make_gurkan.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+import discord
+from bot.bot import Bot
+from bot.constants import Roles
+from discord.ext import commands
+from discord.utils import get
+
+
+class MakeGurkan(commands.Cog):
+    """Make Gurkan Cog adds or removes the gurkan role depending on if the user has gurkan in their name."""
```
Comment (gustavwilliam) [inline bot/exts/gurkan/make_gurkan.py:11]:
```suggestion
    """Makes sure that only members with gurkan names have the gurkan role."""
```
Let's make it a bit more concise.

File: bot/exts/gurkan/make_gurkan.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    def __init__(self, bot: Bot) -> None:
+        self.bot = bot
+
+    @commands.Cog.listener()
+    async def on_member_update(
+        self, before: discord.Member, after: discord.Member
+    ) -> None:
+        """
+        Adds/Removes the role on member update.
```
Comment (gustavwilliam) [inline bot/exts/gurkan/make_gurkan.py:21]:
```suggestion
        Adds/Removes the gurkan role on member update.
```
More specific.

File: bot/exts/gurkan/make_gurkan.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+        When the member updates their nickname or username, this function will check if member is
+        classified to be a gurkan or not.
+        """
+        if after.bot:
+            return
+        if before.display_name != after.display_name:
+            role = get(after.guild.roles, id=Roles.gurkans)
+            if re.search(r"gurk|urkan", str(after.display_name)):
```
Comment (gustavwilliam) [inline bot/exts/gurkan/make_gurkan.py:30]: Let's use the function for checking gurkans, that's already defined in utils.

File: bot/exts/gurkan/make_gurkan.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                await after.add_roles(role)
+            else:
+                await after.remove_roles(role)
```
Comment (gustavwilliam) [inline bot/exts/gurkan/make_gurkan.py:33]: What if they start out with a gurkan nickname and end on a new, but still gurkan, nickname? Will this just be ignored and still work out?

The same thing if they aren't gurkan and end up not being gurkan either.

File: bot/exts/gurkan/make_gurkan.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+    @commands.Cog.listener()
+    async def on_member_join(self, member: discord.Member) -> None:
+        """When the member joins, this function will check if member is classified to be a gurkan or not."""
```
Comment (gustavwilliam) [inline bot/exts/gurkan/make_gurkan.py:37]:
```suggestion
        """Adds the gurkan role to new members who are classified as gurkans."""
```
Concise!

File: bot/exts/gurkan/make_gurkan.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        if member.bot:
+            return
+        role = get(member.guild.roles, id=Roles.gurkans)
+        if re.search(r"gurk|urkan", str(member.name)):
```
Comment (gustavwilliam) [inline bot/exts/gurkan/make_gurkan.py:41]:
```suggestion
        if re.search(r"gurk|urkan", str(member.display_name)):
```
I see no reason to not use the display name. If they've joined before, had a gurkan nickname and are rejoining, will the nickname be left? I don't know, but I'd rather not risk it.

**Review by gustavwilliam — CHANGES_REQUESTED** [review]
Thanks! This looks pretty good, but I have a few comments to resolve.

File: bot/exts/gurkan/make_gurkan.py (hunk 6 above)
Comment (RandomDev26) [reply bot/exts/gurkan/make_gurkan.py:41]: Yea, display name would be better.

**Review by gustavwilliam — APPROVED** [review]
LGTM! Thanks.

I haven't tested this, but the code looks good.

File: Pipfile [hunk 7]
```diff
@@ -7,6 +7,8 @@ verify_ssl = true
 "discord.py" = "~=1.5.1"
 loguru = "~=0.5.3"
 PyYAML = "~=5.3.1"
+win32-setctime = "*"
+flake8 = "*"
```
Comment (vivax3794) [inline Pipfile:11]: why where these added, I dont see them mentioned in the added imports?

Comment (RandomDev26) [reply Pipfile:11]: When running it on windows, it requires that those be installed

Comment (vivax3794) [reply Pipfile:11]: I dont see how any of those 2 are required to run the bot, maybe I am missing something?

Comment (RandomDev26) [reply Pipfile:11]: It required me to install flake8 for linting, and win32-setctime is needed by loguru.

Comment (vivax3794) [reply Pipfile:11]: flake8 would have been installed when you do a dev-install (you see it is listed under dev-dependencies) `pipenv sync --dev`

and I would have assumed `win32-setctime` would have been automatically installed if it is a dependency of `loguru`? but I guess I am wrong.

Comment (RandomDev26) [reply Pipfile:11]: IDK, maybe I messed up while installing all the dependencies. Should I remove it?

Comment (vivax3794) [reply Pipfile:11]: flake8 you can remove, I would assume you could remove `win32-setctime`, but I cant't test that as I am not on windows

Comment (Shivansh-007) [reply Pipfile:11]: Also be specific `*` is not acceptable.

**Review by Shivansh-007 — CHANGES_REQUESTED** [review]

File: Pipfile (hunk 7 above)
Comment (RandomDev26) [reply Pipfile]: Done!

**Comment by RandomDev26** [comment]
I'm bad at pipenvs, but I think I fixed it now

**Review by Shivansh-007 — CHANGES_REQUESTED** [review]
Can you reset all the changes in `Pipfile` and `Pipfile.lock`. They aren't needed.

**Comment by RandomDev26** [comment]
Uhhh... How do I do that?

**Comment by RandomDev26** [comment]
@Shivansh-007 done!

**Review by Shivansh-007 — APPROVED** [review]
Thanks! Looks Good To Me!

**Review by vivax3794 — APPROVED** [review]
LGTM

**Comment by codephile1221** [comment]
Hello there, looks good to me!

**Review by codephile1221 — APPROVED** [review]
Hello there, looks good to me!

**Pull request merged**